### ISSUE TITLE
Performance Regressions

### ISSUE THREAD

**Issue by lukeapage** [dotless/dotless#489]
Against some large private css source

dotless 1.3.0 - 3 seconds
dotless 1.5.0 - 7 seconds
less.js 2.5.1 - 1.6 seconds

Something is wrong here.. C# should be potentially faster than JavaScript.

Ideas:
 - Cache selector toString, used when looking up mixins - saved 10%
 - re-write variable lookup to not use linq - saved 15%

But it still seems too slow..

**Comment by lukeapage** [comment]
Heres what I did for those speedups, some of them make simplications that may not be possible in the code base, I was just trying to work out a best case:

In Ruleset.cs
```
        public Rule Variable(string name, Node startNode)
        {
            Ruleset startNodeRuleset = startNode as Ruleset;

            foreach(var r in Rules.Reverse())
            {
                var rule = r as Rule;
                if (rule == null)
                    continue;

                if (!rule.Variable)
                {
                    continue;
                }
                if (rule.Name == name)
                {
                    return rule;
                }
            }
            return null;
/*
            return Rules
                .TakeWhile(r => r != startNode && (startNodeRuleset == null || !startNodeRuleset.IsEqualOrClonedFrom(r)))
                .OfType<Rule>()
                .Where(r => r.Variable)
                .Reverse()
                .FirstOrDefault(r => r.Name == name);*/

```
and in Node.cs
```
        private string _cachedToCSS = null;
        public virtual string ToCSS(Env env)
        {
            if (_cachedToCSS != null)
                return _cachedToCSS;

            _cachedToCSS = env.Output
                .Push()
                .Append(this)
                .Pop()
                .ToString();

            return _cachedToCSS;
        }
```

**Comment by rytmis** [comment]
Yeah, I was aware that 1.5 came with a performance hit -- I removed a whole bunch of caching, because the caching ended up causing all sorts of bugs in more complex scenarios. I've done some profiling earlier and have some ideas for optimization. Unfortunately, a lot of those ideas require API breakage, so I haven't really started with them yet.

**Comment by UweKeim** [comment]
Any update on this? I would _love_ to have a blazing fast version again.

**Comment by rytmis** [comment]
The performance profiles point to spending a sizable chunk of the processing time within the various mechanisms that combine and evaluate selectors. This code isn't new to dotless 1.5 -- what is, though, is the way we now invoke those operations when doing things like mixin resolution. For instance, running dotless 1.5.2 on the latest Bootstrap 3 sources produces a trace where we call the NodeList<T> constructor well over 300 000 times:

![image](https://cloud.githubusercontent.com/assets/121952/22935143/4bfe6d42-f2da-11e6-83cf-802ceeb3ede6.png)

I don't have the relevant trace from the 1.3 branch yet, but I expect that's one part where you'll see the biggest delta between the versions. The reason is that in 1.5 I made the decision to support mixing in rulesets that have their selectors themselves generated from variables. One such example would be Bootstrap's column classes. This was at least at the time a feature that less.js didn't support, by the way... 

Anyway, in order to make that work, what I did was make dotless more aggressively evaluate selectors. Now, I'm pretty sure we evaluate the same selectors over and over for who knows how many times right now, but since a lot of the code involved happens to be in one of the most convoluted paths of dotless, the correct caching strategy is somewhat non-obvious.

I've been thinking about throwing away the notion of SemVer with regards to dotless' parse tree / compiler API. Were I to do that, I could conceivably start to rework some of the internals in a way that might allow for more obvious caching wins. But then again, I've been thinking about that for a rather long time now... :)
